File: lib/game.go
package lib

import (
	"fmt"
)

// Game ...
type Game struct {
	Deck    Deck
	Board   Cards
	Players Players
	Seed    int64
}

// Join ...
func (game *Game) Join(player *Player) {
	game.Players = append(game.Players, player)
}

// Deal ...
func (game *Game) Deal(cards *Cards) *Card {
	card := game.Deck[0]
	game.Deck = game.Deck[1:]
	*cards = append(*cards, card)
	return card
}

// Init ...
func (game *Game) Init() {
	game.Deck = make(Deck, DeckSize)
	game.Deck.Init()
	game.Deck.Shuffle(game.Seed)
	for i := range game.Players {
		game.Players[i].Game = game
	}
	for j := 0; j < PocketSize; j++ {
		for i := range game.Players {
			game.Deal(&game.Players[i].Pocket)
		}
	}
}

// Step ...
func (game *Game) Step() {
	if len(game.Board) == 0 {
		for i := 0; i < FlopSize; i++ {
			game.Deal(&game.Board)
		}
	} else if len(game.Board) < BoardSize {
		game.Deal(&game.Board)
	}
}

// String ...
func (game *Game) String() string {
	str := "----\n"
	for i := range game.Players {
		str = str + game.Players[i].String() + "\n"
	}
	str = str + fmt.Sprintf("board:%s\n", game.Board.String())
	str = str + fmt.Sprintf("deck:%s\n", Cards(game.Deck).String())
	str = str + fmt.Sprintf("seed:%d\n", game.Seed)
	return str
}

File: lib/Game.js
import lodash from 'lodash'
import Deck from './Deck'
import Player from './Player'
import {POCKET_SIZE, BOARD_SIZE, FLOP_SIZE} from './Const'

export default class Game {
  constructor() {
    this.players = []
    this.board = []
    this.deck = []
  }
  async addPlayer(id) {
    await this.join(new Player(id))
  }
  async join(player) {
    await this.sleep(Math.random() * 100)
    this.players.push(player)
  }
  async init() {
    this.board.length = 0
    this.deck.length = 0
    this.players.forEach(a => a.init(this))
    this.deck.push(...Deck.random())
    for (let i = 0; i < POCKET_SIZE; i++) {
      for (let j = 0; j < this.players.length; j++) {
        await this.deal(this.players[j])
      }
    }
  }
  async deal(player) {
    await this.sleep()
    if (player) {
      player.cards.push(this.deck.pop())
    } else {
      this.board.push(this.deck.pop())
    }
  }
  async flop() {
    await this.sleep()
    if (this.board.length !== 0) {
      throw new Error()
    }
    for (let i = 0; i < FLOP_SIZE; i++) {
      await this.deal()
    }
  }
  async turn() {
    if (this.board.length < FLOP_SIZE || this.board.length >= BOARD_SIZE) {
      throw new Error()
    }
    await this.deal()
  }
  async step() {
    if (this.busy) {
      return true
    } else {
      this.busy = true
    }
    if (this.board.length === 0) {
      await this.flop()
    } else if (this.board.length >= FLOP_SIZE && this.board.length < BOARD_SIZE) {
      await this.turn()
    } else {
      this.end = true
      console.debug("end")
    }
    this.busy = false
    return !this.end
  }
  sleep(time) {
    return new Promise(resolve => setTimeout(resolve, time || 10))
  }
}
